### PR TITLE
FastCV Extension code for Opencv APIs

### PR DESCRIPTION
**Pull request by sssanjee-quic**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/fastcv/test/test_bilateral.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#include "test_precomp.hpp"
+
+namespace opencv_test { namespace {
+
+typedef testing::TestWithParam<tuple<cv::Size,int,int>> fcv_bilateralFilterTest;
+
+TEST_P(fcv_bilateralFilterTest, accuracy)
+{	
+    cv::Size size  = get<0>(GetParam());
+	int d = get<1>(GetParam());
```
Comment (savuor) [inline modules/fastcv/test/test_bilateral.cpp:15]: Here and below: indentation

File: modules/fastcv/test/test_bilateral.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	
+	RNG& rng = cv::theRNG();
+    Mat src(size, CV_8UC1);
+    cvtest::randUni(rng, src, Scalar::all(0), Scalar::all(256));
+
+    cv::Mat dst;
+
+	cv::fastcv::bilateralFilter(src, dst, d, sigmaColor, sigmaSpace);
+	
+    EXPECT_FALSE(dst.empty());
```
Comment (savuor) [inline modules/fastcv/test/test_bilateral.cpp:27]: Not enough checks
We should compare this image to some reference implementation (e.g. from OpenCV) or at least to source image with some fixed threshold, just to ensure that the result resembles smoothed image somehow.

File: modules/fastcv/test/test_moments.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+typedef testing::TestWithParam<tuple<bool,Size,int>> fcv_momentsTest;
+
+TEST_P(fcv_momentsTest, accuracy)
+{
+    const bool binaryImage = get<0>(GetParam());
+    const Size srcSize = get<1>(GetParam());
+    const MatDepth srcType = get<2>(GetParam());
+    Mat src(srcSize, srcType);
+
+	for(int j = 0; j < srcSize.width; ++j)
```
Comment (savuor) [inline modules/fastcv/test/test_moments.cpp:20]: we have `cv::RNG::fill()` for that

File: modules/fastcv/test/test_moments.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+			else if(srcType == CV_32SC1)
+				src.at<int>(i, j) = cv::randu<int>();
+			else if(srcType == CV_32FC1)
+				src.at<float>(i, j) = cv::randu<float>();
+	    }
+
+	cv::Moments m = cv::fastcv::moments(src, binaryImage);
+
+    int len_m = sizeof(m)/sizeof(m.m00);
+    EXPECT_FALSE(len_m != 24);
```
Comment (savuor) [inline modules/fastcv/test/test_moments.cpp:34]: This check will always be true since the size of `cv::Moments` is fixed.
Again, there should be a comparison to some reference implementation or to fixed results.

File: modules/fastcv/test/test_moments.cpp (hunk 3 above)
Comment (savuor) [reply modules/fastcv/test/test_moments.cpp:20]: Optional: since this is the uniform distribution, we can calculate true moments analytically.

File: modules/fastcv/test/test_remap.cpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+    cv::Mat remapOpenCV;
+    cv::remap(src_converted, remapOpenCV, map_x, map_y, interpolation);
+
+    cv::Mat diffImage;
+    cv::absdiff(dst, remapOpenCV, diffImage);
+
+    // Calculate the maximum difference
+    double maxVal=0.0;
+    cv::minMaxLoc(diffImage, nullptr, &maxVal);
```
Comment (savuor) [inline modules/fastcv/test/test_remap.cpp:85]: L_inf norm does the same: `double maxVal = cv::norm(dst, remapOpenCV, cv::NORM_INF);

File: modules/fastcv/test/test_remap.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    // Assert if the difference is acceptable (max difference should be less than 10)
+    CV_Assert(maxVal < 10 && "Difference between images is too high!");
```
Comment (savuor) [inline modules/fastcv/test/test_remap.cpp:116]: The same about L_inf norm

File: modules/fastcv/test/test_scale.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+    cv::Mat diffImage;
+    cv::absdiff(resized_image, resizedImageOpenCV, diffImage);
+
+    // Calculate the maximum difference
+    double maxVal=0.0;
+    cv::minMaxLoc(diffImage, nullptr, &maxVal);
+
+    // Assert if the difference is acceptable (max difference should be less than 10)
+    CV_Assert(maxVal < 10 && "Difference between images is too high!");
```
Comment (savuor) [inline modules/fastcv/test/test_scale.cpp:36]: The same about L_inf norm

File: modules/fastcv/test/test_scale.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
```
Comment (savuor) [inline modules/fastcv/test/test_scale.cpp:61]: The same about L_inf norm

File: modules/fastcv/test/test_scale.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    randu(inputImage, Scalar::all(0), Scalar::all(255)); // Fill with random values
+
+    Size dsize;
+    cv::Mat resized_image;
+
+    // Resize the image by a factor of 2
+    cv::fastcv::resizeDownBy2(inputImage, resized_image);
+
+    // Check if the output size is correct
+    EXPECT_EQ(resized_image.size().width, size.width * 0.5);
```
Comment (savuor) [inline modules/fastcv/test/test_scale.cpp:78]: These checks can be moved to the `accuracy` test

File: modules/fastcv/test/test_scale.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    randu(inputImage, Scalar::all(0), Scalar::all(255)); // Fill with random values
+
+    Size dsize;
+    cv::Mat resized_image;
+
+    // Resize the image by a factor of 2
+    cv::fastcv::resizeDownBy4(inputImage, resized_image);
+
+    // Check if the output size is correct
+    EXPECT_EQ(resized_image.size().width, size.width * 0.25);
```
Comment (savuor) [inline modules/fastcv/test/test_scale.cpp:96]: These checks can be moved to the `accuracy` test

**Review by savuor — COMMENTED** [review]
The rest is OK, waiting for other wrapped functions to be added.

The tests can be merged as-is, without fixing all my comments. But I still have to leave these comments so that they can be fixed in the future.

File: modules/fastcv/test/test_cluster_euclidean.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        std::vector<double> diffs;
+        for (int i = 0; i < nClusters; i++)
+        {
+            double cs = std::abs((trueClusterSizes[i] - clusterSizes[i]) / double(trueClusterSizes[i]));
+            diffs.push_back(cs);
+        }
+        double normL2  = cv::norm(diffs, NORM_L2) / nClusters;
+
+        EXPECT_LT(normL2, 0.392);
```
Comment (asmorkalov) [inline modules/fastcv/test/test_cluster_euclidean.cpp:107]: Average distance hides corner cases, e.g. all points, besides one are the same, but the last one is significant outlier. The check should be reworked.

File: modules/fastcv/test/test_cluster_euclidean.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    {
+        Mat bindings8u, trueBindings8u;
+        Mat(clusterBindings).convertTo(bindings8u, CV_8U);
+        Mat(trueClusterBindings).convertTo(trueBindings8u, CV_8U);
+        double normH = cv::norm(bindings8u, trueBindings8u, NORM_HAMMING) / nPts;
+        EXPECT_LT(normH, 0.658);
```
Comment (asmorkalov) [inline modules/fastcv/test/test_cluster_euclidean.cpp:115]: The same

File: modules/fastcv/test/test_fast10.cpp [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+TEST_P(Fast10Test, accuracy)
+{
+    auto p = GetParam();
+    bool useScores  = std::get<0>(p);
+    int barrier     = std::get<1>(p);
+    int border      = std::get<2>(p);
+    bool nmsEnabled = std::get<3>(p);
+
+    cv::Mat src = imread(cvtest::findDataFile("cv/shared/lena.png"), cv::IMREAD_GRAYSCALE);
```
Comment (asmorkalov) [inline modules/fastcv/test/test_fast10.cpp:21]: Lena note.

File: modules/fastcv/test/test_thresh.cpp [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    int hiThresh = std::get<2>(p);
+    int trueValue  = std::get<3>(p);
+    int falseValue = std::get<4>(p);
+
+    int lowThresh  = std::min(loThresh, hiThresh);
+    int highThresh = std::max(loThresh, hiThresh);
+
+    RNG& rng = cv::theRNG();
+    Mat src(size, CV_8UC1);
+    cvtest::randUni(rng, src, Scalar::all(0), Scalar::all(256));
```
Comment (asmorkalov) [inline modules/fastcv/test/test_thresh.cpp:27]: 256 - overflow.

File: modules/fastcv/src/utils.cpp [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+ * Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+*/
+
+#include "precomp.hpp"
+
+namespace cv {
+namespace fastcv {
+
+bool isInitialized = false;
```
Comment (asmorkalov) [inline modules/fastcv/src/utils.cpp:11]: The same initialization issue as in main repo.

Comment (quic-apreetam) [reply modules/fastcv/src/utils.cpp:11]: will add singleton similar to main

File: modules/fastcv/include/opencv2/remap.hpp [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * @brief Applies a generic geometrical transformation to a 4-channel CV_8UC4 image with bilinear or nearest neighbor interpolation
+ * @param src The first input image data, type CV_8UC4
+ * @param dst The output image data, type CV_8UC4
+ * @param map1 Floating-point CV_32FC1 matrix with each element as the column coordinate of the mapped location in the source image
+ * @param map2 Floating-point CV_32FC1 matrix with each element as the row coordinate of the mapped location in the source image.
+ * @param interpolation Only INTER_NEAREST and INTER_LINEAR interpolation is supported
+*/
+CV_EXPORTS_W void remapRGBA( InputArray src, OutputArray dst,
+                             InputArray map1, InputArray map2, int interpolation);
```
[the remark this answers is not in the document]
Comment (quic-apreetam) [reply modules/fastcv/include/opencv2/remap.hpp:43]: it is to simplify. remapRGBA does not support borderValue but remap does.

**Comment by asmorkalov** [comment]
Closed in favor of https://github.com/opencv/opencv_contrib/pull/3824
